### PR TITLE
Resource requirement updates

### PR DESCRIPTION
**Pull request by atmiguel**

### CODE REVIEW

File: python/README.md [hunk 1]
```diff
@@ -34,7 +34,41 @@ def do_something(params: str):
...
+Additionally, you can specify the memory requirments in either Mebibytes (`MiB`) or Gibibytes (`GiB`).
+
+```python
+from preemo.worker import register
+
+@register(name="some_name", namespace="dev", cpu_requirements={ "cores": 4, "memory": { "MiB": 100 } })
+def do_something(params: bytes):
+    ...
+```
+
```
Comment (atmiguel) [inline python/README.md:62]: todo: I need to add a section explaining how to specify storage requirements

File: python/README.md [hunk 2]
```diff
@@ -34,7 +34,41 @@ def do_something(params: str):
...
+Alternatively, if you need GPUs to effectively run your function, you may pass a different configuration object instead.
+
+```python
+from preemo.worker import register
+
+@register(name="some_name", namespace="dev", gpu_requirements={ "model": "nvidia-tesla-k80" })
```
Comment (atmiguel) [inline python/README.md:68]: After chatting with Forrest, we're going to prevent users from specifying a number of GPU cores at the moment. If I understood correctly, we don't want a user to request an enormous machine while specifying they just need one GPU core.

Comment (atmiguel) [reply python/README.md:68]: @fmmoret, should users be allowed to specify storage requirements for GPUs? Or should we hold off on that too?

Comment (atmiguel) [reply python/README.md:68]: okay, after a chat with Dez, it seems we're going to want to allow GPU core count specification for smaller machines that allow provisioning specific amounts.

I'm going to add the options here, but with the understanding that some combinations of requirements are invalid. We will have a validation step in the scheduler_service that will prevent users from registering a function with invalid requirements.

File: protobuf/endpoints/register_function.proto [hunk 3]
```diff
@@ -4,8 +4,50 @@ package endpoints;
...
+
+  // Optional field indicating amount of storage needed.
+  // Not all precisions may be accepted.
+  // For example, the worker may choose to only accept values that are multiples of Mebibytes.
+  optional uint64 storage_in_bytes = 3;
+}
+
+enum GpuModel {
+  GPU_MODEL_UNSPECIFIED = 0;
+  GPU_MODEL_NVIDIA_TESLA_K80 = 1;
```
Comment (preemoDez) [inline protobuf/endpoints/register_function.proto:26]: It might be a good time, to figure out which models are we going to support. I put some random stuff mostly for testing purposes, but since this is a public proto, we should think about it a little.

Comment (atmiguel) [reply protobuf/endpoints/register_function.proto:26]: ah okay good to know. I'm going to remove this enum for now anyway, but it will be good to document which specific machines are supported. It probably makes sense to just start with a couple (maybe even 1) just as an mvp

File: protobuf/endpoints/register_function.proto [hunk 4]
```diff
@@ -4,8 +4,55 @@ package endpoints;
...
+  // For example, the worker may choose to only accept values that are multiples of Mebibytes.
+  optional uint64 memory_in_bytes = 2;
+
+  // Optional field indicating amount of storage needed.
+  // Not all precisions may be accepted.
+  // For example, the worker may choose to only accept values that are multiples of Mebibytes.
+  optional uint64 storage_in_bytes = 3;
+}
+
+message GpuRequirements {
```
Comment (preemoDez) [inline protobuf/endpoints/register_function.proto:24]: Do we support the jobs requiring GPUs with some computation offloaded to the CPU?

File: protobuf/endpoints/register_function.proto [hunk 5]
```diff
@@ -4,8 +4,55 @@ package endpoints;
...
+  // Required field indicating which GPU machine should be used to execute the function.
+  optional string gpu_model = 1;
```
Comment (preemoDez) [inline protobuf/endpoints/register_function.proto:26]: The comment says `Required`.

File: python/README.md [hunk 6]
```diff
@@ -34,7 +34,41 @@ def do_something(params: str):
...
+def do_something(params: bytes):
+    ...
+```
+
+If you need GPUs to effectively run your function, you may pass the specific model to use.
+
+```python
+from preemo.worker import register
+
+@register(name="some_name", namespace="dev", gpu="nvidia-tesla-k80")
```
Comment (preemoDez) [inline python/README.md:68]: Should we verify the passed GPU model? Does Google Cloud provide an instance with k80 GPU?

File: protobuf/endpoints/register_function.proto (hunk 5 above)
Comment (atmiguel) [reply protobuf/endpoints/register_function.proto:26]: Yeah, proto3 doesn't actually allow you to make a field required. Maybe it's the wrong move, but I'm trying to leave comments on most fields to indicate whether or not they are actually optional. Or in other words, I'm trying to indicate if the server will expect their presence and return an error if they're missing.

File: python/README.md (hunk 6 above)
Comment (atmiguel) [reply python/README.md:68]: The verification will all happen in the scheduler service

**Review by preemoDez — APPROVED** [review]

File: python/preemo/worker/_worker_client.py [hunk 7]
```diff
@@ -91,6 +97,74 @@ def __call__(self, params: Optional[bytes] = None) -> Optional[Result]:
 
 # TODO(adrian@preemo.io, 05/11/2023): add logging
 class WorkerClient:
+    @staticmethod
+    def _ensure_value_is_non_negative(
+        *, name: str, value: Optional[Union[int, float]]
+    ) -> None:
+        if value is not None and value < 0:
+            raise Exception(f"{name} must not be negative")
```
Comment (lilwilsond2) [inline python/preemo/worker/_worker_client.py:105]: what does this have to do with a worker client

Comment (atmiguel) [reply python/preemo/worker/_worker_client.py:105]: nothing. I can move it

File: python/preemo/worker/_bytes.py [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    # We have to cast here because mypy will not implement type narrowing for TypedDict
+    # See https://github.com/python/mypy/issues/11080
+
+    if mebibyte_literal in byte_dict:
+        return convert_mebibytes_to_bytes(cast(MebibyteDict, byte_dict))
+
+    if gibibyte_literal in byte_dict:
+        return convert_gibibytes_to_bytes(cast(GibibyteDict, byte_dict))
+
+    raise Exception(f"byte_dict had unexpected keys: {byte_dict}")
```
Comment (lilwilsond2) [inline python/preemo/worker/_bytes.py:40]: if this can't have both keys anyway, does just having a `Mebibyte` and `Gibibyte` class that fulfill some interface with a `getByteSize()` make more sense

Comment (atmiguel) [reply python/preemo/worker/_bytes.py:40]: They're both extending `TypedDict` which are kinda finicky. I can try creating a `Protocol` and having them both extend that protocol. I'm not convinced it'll work, but I'll give it a shot

Comment (lilwilsond2) [reply python/preemo/worker/_bytes.py:40]: i was thinking you could drop the `typeddict`ness

Comment (atmiguel) [reply python/preemo/worker/_bytes.py:40]: I'm not sure how to do that without changing the interface for the user.  There are certainly other user interfaces that could work just as well, but that's a separate discussion. I think to get the user interface we currently have designed, it needs to be a TypedDict

File: python/README.md (hunk 6 above)
Comment (preemoDez) [reply python/README.md:68]: How does the user know what kind of GPUs we support? Is there a way to let them know, without duplicating the code? For example, sent the enum values to the client and render it somewhere.

Comment (atmiguel) [reply python/README.md:68]: I plan on making a python enum with all the options, for ease of use

Comment (atmiguel) [reply python/README.md]: added a todo. Will take care of this in a subsequent PR

File: protobuf/endpoints/register_function.proto [hunk 9]
```diff
@@ -4,8 +4,55 @@ package endpoints;
...
+
+  // Optional field indicating the number of GPU cores needed.
+  optional uint32 cores = 2;
```
[the remark this answers is not in the document]
Comment (atmiguel) [reply protobuf/endpoints/register_function.proto]: changed to `gpu_count`

File: protobuf/endpoints/register_function.proto [hunk 10]
```diff
@@ -4,8 +4,55 @@ package endpoints;
...
+    GpuRequirements gpu = 2;
+  }
+}
+
 message RegisterFunctionRequest {
+  // Required.
   optional models.RegisteredFunction function_to_register = 1;
+
+  // Optional field indicating resources that should be allocated for executing this function.
+  optional ResourceRequirements resource_requirements = 2;
```
Comment (lilwilsond2) [inline protobuf/endpoints/register_function.proto:55]: relevant more places than just here but do we have a metric to determine when a field is "optional" (jfc that's overloaded) vs when we do something like [Value (which might deserve a more descriptive name/path at this point)](https://github.com/Preemo-Inc/worker-sdk/blob/f2344c9f2f5ceb47cb48ec8be0d4719334c738b0/protobuf/models/value.proto#L7)

File: python/README.md [hunk 11]
```diff
@@ -34,7 +34,53 @@ def do_something(params: str):
...
+#### Resource Requirements
+
+If your function has particular resource requirements, you should specify those during registration.
+
+For example, if your function needs 4 cpu cores to run effectively, you can pass a configuration object.
+
+```python
+from preemo.worker import register
+
+@register(name="some_name", namespace="dev", cpu_cores=4)
```
Comment (lilwilsond2) [inline python/README.md:48]: does it make more sense to have the argument be millicores so there's no question of what precision is supported

i would say you could still have a `cpu_cores` arg as an `int` for a bit of a sugar but then you have to get into function overloading or special casing around only one of the two args should be present 🤮

**Review by lilwilsond2 — APPROVED** [review]

File: python/README.md (hunk 11 above)
Comment (atmiguel) [reply python/README.md:48]: it's something we can consider adding later if people run into it. I expect no one will intentionally ask for `3.0001` cores, but we'll see lol

File: protobuf/endpoints/register_function.proto (hunk 10 above)
Comment (atmiguel) [reply protobuf/endpoints/register_function.proto:55]: okay understood, this is about the different ways of indicating a value's presence. In Value, we use a null_value, but here we're using optional. I think the pattern we should learn towards is optional, since proto forces us to make everything optional anyway. I'll leave a todo on the Value proto to move it towards an optional pattern
